File: features/transaction/purchase_platform/src/main/java/com/tokopedia/purchase_platform/common/di/module/MultipleAddressModule.java
package com.tokopedia.purchase_platform.common.di.module;

import com.tokopedia.purchase_platform.features.checkout.domain.usecase.ChangeShippingAddressUseCase;
import com.tokopedia.purchase_platform.features.checkout.subfeature.multiple_address.domain.usecase.GetCartMultipleAddressListUseCase;
import com.tokopedia.purchase_platform.common.di.scope.MultipleAddressScope;
import com.tokopedia.purchase_platform.features.checkout.subfeature.multiple_address.view.IMultipleAddressPresenter;
import com.tokopedia.purchase_platform.features.checkout.subfeature.multiple_address.view.IMultipleAddressView;
import com.tokopedia.purchase_platform.features.checkout.subfeature.multiple_address.view.MultipleAddressPresenter;
import com.tokopedia.purchase_platform.common.data.repository.ICartRepository;
import com.tokopedia.purchase_platform.common.utils.CartApiRequestParamGenerator;
import com.tokopedia.user.session.UserSession;
import com.tokopedia.user.session.UserSessionInterface;

import dagger.Module;
import dagger.Provides;
import rx.subscriptions.CompositeSubscription;

/**
 * Created by kris on 2/5/18. Tokopedia
 */

@Module(includes = TrackingAnalyticsModule.class)
public class MultipleAddressModule {

    private final IMultipleAddressView view;

    public MultipleAddressModule(IMultipleAddressView view) {
        this.view = view;
    }

    @MultipleAddressScope
    @Provides
    CompositeSubscription provideCompositeSubscription() {
        return new CompositeSubscription();
    }

//    @MultipleAddressScope
//    @Provides
//    ChangeShippingAddressUseCase provideMultipleAddressUseCase(ICartRepository repository) {
//        return new ChangeShippingAddressUseCase(repository);
//    }

    @MultipleAddressScope
    @Provides
    UserSessionInterface provideUserSessionInterface() {
        return new UserSession(view.getActivityContext());
    }

    @MultipleAddressScope
    @Provides
    IMultipleAddressPresenter providePresenter(ChangeShippingAddressUseCase changeShippingAddressUseCase,
                                               GetCartMultipleAddressListUseCase getCartMultipleAddressListUseCase,
                                               CartApiRequestParamGenerator cartApiRequestParamGenerator,
                                               UserSessionInterface userSessionInterface) {
        return new MultipleAddressPresenter(getCartMultipleAddressListUseCase,
                changeShippingAddressUseCase, cartApiRequestParamGenerator, userSessionInterface);
    }
}

File: features/transaction/purchase_platform/src/main/java/com/tokopedia/purchase_platform/features/checkout/subfeature/multiple_address/view/di/MultipleAddressModule.kt
package com.tokopedia.purchase_platform.features.checkout.subfeature.multiple_address.view.di

import com.tokopedia.purchase_platform.common.di.PurchasePlatformBaseModule
import com.tokopedia.purchase_platform.common.di.PurchasePlatformCommonModule
import com.tokopedia.purchase_platform.common.di.PurchasePlatformNetworkModule
import com.tokopedia.purchase_platform.common.di.PurchasePlatformQualifier
import com.tokopedia.purchase_platform.common.utils.CartApiRequestParamGenerator
import com.tokopedia.purchase_platform.features.cart.domain.mapper.CartMapper
import com.tokopedia.purchase_platform.features.cart.domain.mapper.ICartMapper
import com.tokopedia.purchase_platform.features.checkout.domain.usecase.ChangeShippingAddressUseCase
import com.tokopedia.purchase_platform.features.checkout.subfeature.multiple_address.data.api.MultipleAddressApi
import com.tokopedia.purchase_platform.features.checkout.subfeature.multiple_address.data.repository.IMultipleAddressRepository
import com.tokopedia.purchase_platform.features.checkout.subfeature.multiple_address.data.repository.MultipleAddressRepository
import com.tokopedia.purchase_platform.features.checkout.subfeature.multiple_address.domain.usecase.GetCartMultipleAddressListUseCase
import com.tokopedia.purchase_platform.features.checkout.subfeature.multiple_address.view.IMultipleAddressPresenter
import com.tokopedia.purchase_platform.features.checkout.subfeature.multiple_address.view.MultipleAddressPresenter
import com.tokopedia.user.session.UserSessionInterface
import dagger.Module
import dagger.Provides
import retrofit2.Retrofit

/**
 * Created by Irfan Khoirul on 2019-08-29.
 */

@Module(includes = [
    PurchasePlatformNetworkModule::class,
    PurchasePlatformBaseModule::class,
    PurchasePlatformCommonModule::class
])
class MultipleAddressModule {

    @Provides
    @MultipleAddressScope
    fun provideMultipleAddressApi(@PurchasePlatformQualifier retrofit: Retrofit): MultipleAddressApi {
        return retrofit.create(MultipleAddressApi::class.java)
    }

    @Provides
    @MultipleAddressScope
    fun provideIMultipleAddressRepository(multipleAddressApi: MultipleAddressApi): IMultipleAddressRepository {
        return MultipleAddressRepository(multipleAddressApi)
    }

    @Provides
    @MultipleAddressScope
    fun provideICartMapper(): ICartMapper {
        return CartMapper()
    }

    @Provides
    @MultipleAddressScope
    fun providePresenter(changeShippingAddressUseCase: ChangeShippingAddressUseCase,
                         getCartMultipleAddressListUseCase: GetCartMultipleAddressListUseCase,
                         cartApiRequestParamGenerator: CartApiRequestParamGenerator,
                         userSessionInterface: UserSessionInterface): IMultipleAddressPresenter {
        return MultipleAddressPresenter(
                getCartMultipleAddressListUseCase,
                changeShippingAddressUseCase,
                cartApiRequestParamGenerator,
                userSessionInterface
        )
    }

}